### PR TITLE
Use the same filter chain factory for plain and TLS connections

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Bugfix
- Documentation

### Description

Fixes: #883. Use the same filter chain factory for plain and TLS connections.

### Additional Context

There are potential use cases where the distinction between plain and TLS connections could be interesting to filter authors but exposing that distinction through the FilteChainFactory instance is not how we should address that (we would add it too `FilterFactoryContext`).

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [x] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterFactory.java [hunk 1]
```diff
@@ -23,7 +23,7 @@ public interface FilterFactory<C, I> {
 
     /**
      * Initializes the factory with the specified configuration.
-     * This method is guaranteed to be called at most once, and before any call to
+     * This method is guaranteed to be called at most once, per factory instance, and before any call to
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterFactory.java:26]: Thinking about this further based on the tests I ended up writing I think this is still not quite right.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/InvocationCountingFilterFactory.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        return config;
+    }
+
+    @NonNull
+    @Override
+    public Filter createFilter(FilterFactoryContext context, Config initializationData) {
+        return (RequestFilter) (apiKey, header, request, requestFilterContext) -> requestFilterContext.forwardRequest(header, request);
+    }
+
+    //    @JsonIgnore
```
Comment (SamBarker) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/InvocationCountingFilterFactory.java:42]: Doh. Need to remove that.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java [hunk 3]
```diff
@@ -79,71 +75,81 @@ public void initChannel(SocketChannel ch) {
...
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private void initPlainChannel(SocketChannel ch, ChannelPipeline pipeline, Optional<String> bindingAddress, int targetPort) {
+        pipeline.addLast(new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) {
+                virtualClusterBindingResolver.resolve(Endpoint.createEndpoint(bindingAddress, targetPort, tls), null)
+                        .handle((binding, t) -> {
                             if (t != null) {
-                                promise.setFailure(t);
+                                ctx.fireExceptionCaught(t);
```
Comment (tombentley) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java:93]: Why was this necessary (I'm not say it's wrong, just interested)

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java:93]: I don't remember changing that. I'm more than a little suspicious the diff is lying(well getting confused) but I will dig further once I finish bashing at a unit test to generate some coverage.

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java:93]: There is a reverse diff around line [129/120](https://github.com/kroxylicious/kroxylicious/pull/905/files/bd12dbbd5f7e26578cc76e4e5528d7b464e7e795#diff-90761a096b3920e7984e680dae477133069134d19d4b01297f42e528531e26c7L129)

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java:93]: The diff I see in IntelliJ doesn't have this change. It shows two big blocks of change rather than small line by line changes.

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyInitializerTest.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class KafkaProxyInitializerTest {
```
Comment (SamBarker) [inline kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyInitializerTest.java:52]: Hmm this is proving to be a big distraction from the point of the PR. I wonder if I should try and move it to a new PR (it only exists to try and pass the coverage gate)

Comment (SamBarker) [reply kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyInitializerTest.java]: Raised #908 to extract this out and reduce the noise.

**Comment by franvila** [comment]
> > [0.0% Coverage on New Code](https://sonarcloud.io/component_measures?id=kroxylicious_kroxylicious&pullRequest=905&metric=new_coverage&view=list) (required ≥ 76%)
> 
> _sigh_ integration tests don't count for coverage...

Is this related to disabling integration tests run on Sonar build? Did it count in previous runs? Just curious

**Review by k-wall — APPROVED** [review]
lgtm, thanks @SamBarker

**Review by k-wall — APPROVED** [review]

**Comment by SamBarker** [comment]
This issue is mostly about wiring so is quite difficult to unit test directly. I started trying to add a unit test but it turned into a [massive diff](https://github.com/kroxylicious/kroxylicious/pull/908) that swamped this change. So we will merge this despite the sonar warning and continue testing efforts in #908
